### PR TITLE
Initial resources

### PR DESCRIPTION
**Pull request by salvianreynaldi**
- Application binary bucket (appbin)
- CodePipeline artifact bucket
- CodeBuild cache bucket
- CodePipeline role
- CodeBuild role
- CloudTrail for the appbin bucket write actions
- Events role (to trigger codepipeline on appbin S3 push)
- Lambda role (to share AMI to other AWS accounts)

### CODE REVIEW

**Comment by salvianreynaldi** [comment]
cc @febryantonius @andysaputra

File: data.tf [hunk 1]
```diff
@@ -0,0 +1,543 @@
...
+    resources = [
+      # these resources might need to be more 'locked down'
+      "arn:aws:ec2:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:key-pair/packer_*",
+
+      "arn:aws:ec2:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:launch-template/*",
+      "arn:aws:ec2:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:network-interface/*",
+      "arn:aws:ec2:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:placement-group/*",
+      "arn:aws:ec2:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:security-group/${aws_security_group.template.id}",
+      "arn:aws:ec2:${data.aws_region.current.name}::snapshot/*",
+      "arn:aws:ec2:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:subnet/*",
```
Comment (rizkidoank) [inline data.tf]: is there any reason why not lock it to app subnet instead of `*`?

File: data.tf [hunk 2]
```diff
@@ -554,3 +554,27 @@ data "aws_iam_policy_document" "lambda_ami" {
...
+
+  statement {
+    effect = "Allow"
+
+    actions = [
+      "lambda:InvokeFunction",
+    ]
+
+    resources = [
+      "arn:aws:lambda:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:function/${var.product_domain}*",
```
Comment (rizkidoank) [inline data.tf]: what is it for? What are the lambda functions that we will provides?

File: examples/beitest-bake-ami/main.tf [hunk 3]
```diff
@@ -0,0 +1,222 @@
+provider "aws" {
+  region = "ap-southeast-1"
+}
+
+module "bei" {
+  source         = "../.."
+  product_domain = "bei"
```
Comment (rafikurnia) [inline examples/beitest-bake-ami/main.tf]:
```bash
Error: module "bei": missing required argument "base_ami_owners"
Error: module "bei": missing required argument "vpc_id"
Error: module "bei": missing required argument "target_accounts"
```

File: examples/beitest-bake-ami/main.tf [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+      version          = "1"
+
+      configuration {
+        ProjectName = "${aws_codebuild_project.bake_ami.name}"
+      }
+
+      run_order = 1
+    }
+  }
+}
```
Comment (rafikurnia) [inline examples/beitest-bake-ami/main.tf]: how about third stage which invoke lambda?

File: examples/beitest-bake-ami/main.tf [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+    "Name"          = "${var.service_name}-bake-ami"
+    "Description"   = "Bake ${var.service_name} AMI"
+    "Service"       = "${var.service_name}"
+    "ProductDomain" = "${var.product_domain}"
+    "Environment"   = "management"
+    "ManagedBy"     = "Terraform"
+  }
+}
+
+resource "aws_lambda_function" "share_ami" {
```
Comment (rafikurnia) [inline examples/beitest-bake-ami/main.tf]: one lambda per service?

File: examples/beitest-bake-ami/main.tf [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+    base-ami-owners           = "${join(",", var.base_ami_owners)}"
+    subnet-id                 = "${join(",", random_shuffle.subnet.result)}"
+    vpc-id                    = "${module.bei.vpc_id}"
+    region                    = "${data.aws_region.current.name}"
+  }
+}
+
+resource "aws_cloudwatch_event_rule" "this" {
+  name        = "trigger-${aws_codepipeline.ami_baking.name}"
+  description = "Capture each s3://${module.bei.application_binary_bucket}/${var.playbook_key} upload"
```
Comment (rafikurnia) [inline examples/beitest-bake-ami/main.tf]: I thought we are agreed to use `traveloka-builds` bucket for now
cc: @rizkidoank

File: main.tf [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+}
+
+module "lambda_function_name" {
+  source = "github.com/traveloka/terraform-aws-resource-naming.git?ref=v0.7.1"
+
+  name_prefix   = "${var.product_domain}-ami-sharing"
+  resource_type = "lambda_function"
+}
+
+resource "aws_lambda_function" "share_ami" {
```
Comment (rafikurnia) [inline main.tf]: where is [aws_lambda_permission](https://www.terraform.io/docs/providers/aws/r/lambda_permission.html) resource?

Comment (rafikurnia) [inline main.tf]: if you write it here, then is it going to be one per product domain?

File: main.tf [hunk 8]
```diff
@@ -0,0 +1,359 @@
...
+
+  name_prefix   = "${var.product_domain}-appbin-${data.aws_caller_identity.current.account_id}-"
+  resource_type = "s3_bucket"
+}
+
+resource "aws_s3_bucket" "application_binary" {
+  bucket = "${module.application_binary.name}"
+  acl    = "private"
+
+  policy = <<POLICY
```
Comment (rafikurnia) [inline main.tf]: I thought we are agreed to use `traveloka-builds` bucket for now
cc: @rizkidoank 

then we should allow put access from jenkins too

File: main.tf [hunk 9]
```diff
@@ -0,0 +1,359 @@
...
+  source_code_hash = "${data.archive_file.share_ami_function.output_base64sha256}"
+  role             = "${module.lambda_role.role_arn}"
+  function_name    = "${module.lambda_function_name.name}"
+  description      = "share ${var.product_domain}'s AMIs"
+  runtime          = "python3.6"
+  handler          = "main.handler"
+}
+
+resource "aws_ssm_parameter" "target_accounts" {
+  name  = "ami-target-accounts"
```
Comment (rafikurnia) [inline main.tf]: if you write it here, then is it going to be one per product domain? if yes then it should have `product_domain` incorporated in its name

File: main.py [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+from pprint import pprint
+
+ec2_client = boto3.client('ec2')
+ssm_client = boto3.client('ssm')
+
+def get_acount_ids():
+    account_ids = []
+    try:
+        # TODO
+        account_ids = ssm_client.get_parameter(Name="ami-target-accounts")["Parameter"]["Value"]
```
Comment (rafikurnia) [inline main.py]: this one too, should access from different path per product domain, if you are willing to create this lambda per product domain

File: data.tf [hunk 11]
```diff
@@ -0,0 +1,620 @@
...
+    effect = "Allow"
+
+    actions = [
+      "s3:GetObject",
+    ]
+
+    resources = [
+      "${aws_s3_bucket.codebuild_cache.arn}/*",
+      "${aws_s3_bucket.codepipeline_artifact.arn}/*",
+      "${aws_s3_bucket.application_binary.arn}/*",
```
Comment (rafikurnia) [inline data.tf]: not using traveloka-builds?

File: examples/beitest-bake-ami/main.tf [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+  filename         = "${data.archive_file.share_ami_function.output_path}"
+  source_code_hash = "${data.archive_file.share_ami_function.output_base64sha256}"
+  role             = "${module.bei.lambda_role_arn}"
+  function_name    = "${var.service_name}-share-ami"
+  description      = "share-${var.service_name}-amis-to-dev-and-prod-account"
+  runtime          = "python3.6"
+  handler          = "main.lambda_handler"
+}
+
+data "archive_file" "share_ami_function" {
```
Comment (rafikurnia) [inline examples/beitest-bake-ami/main.tf]: why is it also here?

File: data.tf (hunk 11 above)
Comment (rafikurnia) [reply data.tf]: based on offline discussion, this comment is not relevant anymore

File: main.tf (hunk 8 above)
Comment (rafikurnia) [reply main.tf]: based on offline discussion, this comment is not relevant anymore

**Pull request merged**